### ISSUE TITLE
AllTooWell - Admin can affect the odds of a raffle in some case

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-08-winnables-raffles-judging#273]
AllTooWell

Medium

# Admin can affect the odds of a raffle in some case

### Summary

Admin can set `maxHoldings = maxTicketSupply` in raffle and players who buy all tickets is 100% guaranteed to win.

### Root Cause

Contest readme:
> The principles that must always remain true are:
Admins cannot affect the odds of a raffle

In `createRaffle` function, `maxHoldings` can be set equal to `maxTicketSupply`.
For example in [testnet competetion 18](https://testnet.winnables.com/competitions/18), `maxHoldings` is equal to `maxTicketSupply`.
https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L252-L274

### Internal pre-conditions

1. Raffle's `maxHoldings` is equal to `maxTicketSupply`.

### External pre-conditions

1. A player have enough AVAX to buy all tickets.

### Attack Path

1. This player call `buyTickets`. The input of `ticketCount` is equal to `maxTicketSupply`.

### Impact

1. This player is 100% guaranteed to win.
2. Admin impacts odds of raffle by setting inappropriate paramater.

### Mitigation

```diff
    function createRaffle(
        uint256 raffleId,
        uint64 startsAt,
        uint64 endsAt,
        uint32 minTickets,
        uint32 maxTickets,
        uint32 maxHoldings
    ) external onlyRole(0) {
        _checkRaffleTimings(startsAt, endsAt);
        if (maxTickets == 0) revert RaffleRequiresTicketSupplyCap();
        if (maxHoldings == 0) revert RaffleRequiresMaxHoldings();
+       require(maxHoldings < maxTickets);
        Raffle storage raffle = _raffles[raffleId];
        if (raffle.status != RaffleStatus.PRIZE_LOCKED) revert PrizeNotLocked();

        raffle.status = RaffleStatus.IDLE;
        raffle.startsAt = startsAt;
        raffle.endsAt = endsAt;
        raffle.minTicketsThreshold = minTickets;
        raffle.maxTicketSupply = maxTickets;
        raffle.maxHoldings = maxHoldings;

        emit NewRaffle(raffleId);
    }
```

**Comment by IAm0x996** [comment]
Escalate

> The principles that must always remain true are:
> Admins cannot affect the odds of a raffle


According to the [Contest readme](https://github.com/sherlock-audit/2024-08-winnables-raffles?tab=readme-ov-file#q-please-discuss-any-design-choices-you-made), this issue should be deemed valid.

**Comment by sherlock-admin3** [comment]
> Escalate
> 
> > The principles that must always remain true are:
> > Admins cannot affect the odds of a raffle
> 
> 
> According to the [Contest readme](https://github.com/sherlock-audit/2024-08-winnables-raffles?tab=readme-ov-file#q-please-discuss-any-design-choices-you-made), this issue should be deemed valid.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by mystery0x** [comment]
All admin related issues have been discussed with the HOJ and the sponsor who wished the following governing statements could have been more carefully written and/or omitted:

"The protocol working as expected relies on having an admin creating raffles. It should be expected that the admin will do their job. However it is not expected that the admin can steal funds that should have ended in a raffle participant’s wallet in any conceivable way."

Much as many of the findings suggest good fixes, the conclusion is that they will all be deemed low unless it's beyond the trusted admin control.

**Comment by WangSecurity** [comment]
The escalation will be accepted, the issue will be validated with medium severity and duplicated with #129. For details, look at the escalation under #129.

**Comment by Brivan-26** [comment]
@WangSecurity This is not a dup of #129  and is an invalid issue.

The summary says:
> Admin can set maxHoldings = maxTicketSupply in raffle and players who buy all tickets is 100% guaranteed to win.

External pre-condition says:
> A player have enough AVAX to buy all tickets.

If a user is rich enough to buy all the tickets supply, it is totally expected (by math and by logic) that he wins the raffle

**Comment by WangSecurity** [comment]
Thank you @Brivan-26 excuse me for this. This issue is indeed invalid and admin doesn't affect anything here. Planning to reject the escalation and leave this issue as it is.

**Comment by IAm0x996** [comment]
> Thank you @Brivan-26 excuse me for this. This issue is indeed invalid and admin doesn't affect anything here. Planning to reject the escalation and leave this issue as it is.

@WangSecurity 6666666

> If a user is rich enough to buy all the tickets supply, it is totally expected (by math and by logic) that he wins the raffle

Since `maxHoldings = maxTicketSupply`, it allows one person to purchase all the tickets and monopolize the rewards. For example, if the Admin sets `maxHoldings = maxTicketSupply = 10` and the cost is less than the reward, then that user could gain more benefits at a very low cost.

Additionally, in a lottery event, there should be at least two participants! However, with `maxHoldings = maxTicketSupply`, it could result in a single person participating and claiming all the rewards. Is this reasonable?

In summary, isn’t this a significant issue?

**Comment by IAm0x996** [comment]
> If a user is rich enough to buy all the tickets supply, it is totally expected (by math and by logic) that he wins the raffle

If a user is wealthy enough to buy all the tickets and there is only one participant in the lottery, do you think there is no impact? Could this lead to money laundering or malicious admin manipulation of rewards, intentionally channeling benefits?

**Comment by IAm0x996** [comment]
Hello @WangSecurity 

In the [summary comment](https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/129#issuecomment-2350690999), I see that you added #273. To avoid confusion in the future, at [10:42](https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/273#issuecomment-2351936055), you considered it invalid. At [10:51](https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/129#issuecomment-2351940789), referring to the latest comments, you considered it valid.

> Planning to reject the escalation and leave this issue as it is.

Therefore, the escalation to this issue should be approved.

**Comment by WangSecurity** [comment]
Thank you for pointing it out. Corrected the duplication list under #129.

By the following line from my comment, I meant that @Brivan-26 was correct that #273 shouldn't be a valid duplicate:
> Also, correct about https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/273.

I stand by my previous decision here, the admin doesn't affect the odds of the raffle and if the user buys all the tickets and wins the raffle, it's not an issue, but how it's supposed to work. Hence, this should remain invalid, planning to reject the escalation.

**Comment by IAm0x996** [comment]
> Thank you for pointing it out. Corrected the duplication list under #129.
> 
> By the following line from my comment, I meant that @Brivan-26 was correct that #273 shouldn't be a valid duplicate:
> 
> > Also, correct about #273.
> 
> I stand by my previous decision here, the admin doesn't affect the odds of the raffle and if the user buys all the tickets and wins the raffle, it's not an issue, but how it's supposed to work. Hence, this should remain invalid, planning to reject the escalation.


Hello @WangSecurity  , This should be valid.

Since maxHoldings = maxTicketsupply , it allows one person to purchase all the tickets and monopoize the rewards, For example, ifthe Admin sets maxoldings = maxTicketsupply = 10 and the cost is less than the reward, then that user could gain more benefitsat a very low cost.

Additionaly, in a lottery event, there should be at least two participants! However, with maxHoldings = maxTicketsupply , it couldresult in a single person participating and claiming all the rewards. ls this reasonable?

> If a user is rich enough to buy all the tickets supply, it is totally expected (by math and by logic) that he wins the raffle

If a user is wealthy enough to buy all the tickets and there is only one participant in the lottery, do you think there is no impact?

could this lead to money laundering or malicious admin manipulation of rewards, intentionally channeling benefits?

In summary, isn't this a sianificant issue?

**Comment by WangSecurity** [comment]
> Since maxHoldings = maxTicketsupply , it allows one person to purchase all the tickets and monopoize the rewards, For example, ifthe Admin sets maxoldings = maxTicketsupply = 10 and the cost is less than the reward, then that user could gain more benefitsat a very low cost.

That doesn't make the admin affect odds of the raffle. Regardless of these values, the user could buy as many tickets as they can to increase their chance of winning, but that doesn't mean the admin affects the odds of the raffle. That's how the protocol is supposed to work and setting maxoldings = maxTicketsupply doesn't affect the odds of the raffle by itself.

> If a user is wealthy enough to buy all the tickets and there is only one participant in the lottery, do you think there is no impact?

That's not the admin affecting the odds of the raffle. Moreover, in that case there is no loss to anyone and there's no benefit for the user doing it, since they just receive their funds back.

> could this lead to money laundering or malicious admin manipulation of rewards, intentionally channeling benefits?

This question is irrelevant, since the repot doesn't show any manipulation of rewards, money laundering or intentionally channeling benefits. Setting maxoldings = maxTicketsupply doesn't show anything of this.

Hence, the decision remains, reject the escalation and leave the issue as it is.

**Comment by IAm0x996** [comment]
> > Since maxHoldings = maxTicketsupply , it allows one person to purchase all the tickets and monopoize the rewards, For example, ifthe Admin sets maxoldings = maxTicketsupply = 10 and the cost is less than the reward, then that user could gain more benefitsat a very low cost.
> 
> That doesn't make the admin affect odds of the raffle. Regardless of these values, the user could buy as many tickets as they can to increase their chance of winning, but that doesn't mean the admin affects the odds of the raffle. That's how the protocol is supposed to work and setting maxoldings = maxTicketsupply doesn't affect the odds of the raffle by itself.
> 
> > If a user is wealthy enough to buy all the tickets and there is only one participant in the lottery, do you think there is no impact?
> 
> That's not the admin affecting the odds of the raffle. Moreover, in that case there is no loss to anyone and there's no benefit for the user doing it, since they just receive their funds back.
> 
> > could this lead to money laundering or malicious admin manipulation of rewards, intentionally channeling benefits?
> 
> This question is irrelevant, since the repot doesn't show any manipulation of rewards, money laundering or intentionally channeling benefits. Setting maxoldings = maxTicketsupply doesn't show anything of this.
> 
> Hence, the decision remains, reject the escalation and leave the issue as it is.

Okay

**Comment by WangSecurity** [comment]
Result:
Invalid
Unique

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [AllTooWell](https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/273/#issuecomment-2330432968): rejected
